drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+	if (config->mode.tx_lsb_first) {
+		v |= I2C_TX_LSB_FIRST;
+	} else {
+		v &= ~I2C_TX_LSB_FIRST;
+	}
+
+	if (config->mode.rx_lsb_first) {
+		v |= I2C_RX_LSB_FIRST;
+	} else {
+		v &= ~I2C_RX_LSB_FIRST;
+	}
+
+	if (dev_config.bits.is_master_device) {
+		/* Act as master. */
+		v |= I2C_MS_MODE;
+
+		*config->regs.slave_addr = 0;
+	} else {
+		u32_t addr;
+
+		/* Act as slave. */
+		v &= ~I2C_MS_MODE;
+
+		addr = data->address & I2C_SLAVE_ADDR_V;
+		*config->regs.slave_addr |= addr << I2C_SLAVE_ADDR_S;
+		if (dev_config.bits.use_10_bit_addr) {
+			*config->regs.slave_addr |= I2C_ADDR_10BIT_EN;
+		}
+
+		/* Before setting up FIFO and interrupts, stop transmission */
+		*config->regs.ctr &= ~I2C_TRANS_START;
+
+		/* Byte after address isn't the offset address in slave RAM */
+		*config->regs.fifo_cfg &= ~I2C_FIFO_ADDR_CFG_EN;
+	}
+
+	/* Use open-drain for clock and data pins */
+	v |= (I2C_SCL_FORCE_OUT | I2C_SDA_FORCE_OUT);
+
+	v |= I2C_CLK_EN;
+
+	/* Sample data when SCL is high */
+	v &= ~I2C_SAMPLE_SCL_LEVEL;
+
+	*config->regs.ctr = v;
+
+	switch (dev_config.bits.speed) {
+	case I2C_SPEED_STANDARD:
+		i2c_esp32_configure_frequency(config, 100000);
+		break;
+	case I2C_SPEED_FAST:
+		i2c_esp32_configure_frequency(config, 400000);
+		break;
+	default:
+		r = -EINVAL;
+		goto out;
+	}
+
+	/* Use FIFO to transmit data */
+	*config->regs.fifo_cfg &= ~I2C_NONFIFO_EN;
+
+	v = CONFIG_I2C_ESP32_TIMEOUT & I2C_TIME_OUT_REG;
+	*config->regs.timeout = v << I2C_TIME_OUT_REG_S;
+
+	/* Enable interrupt types handled by the ISR */
+	*config->regs.intr |= I2C_ACK_ERR_INT_ENA_M |
+			      I2C_TIME_OUT_INT_ENA_M |
+			      I2C_TRANS_COMPLETE_INT_ENA_M |
+			      I2C_ARBITRATION_LOST_INT_ENA_M;
+
+	/* Activate the device IRQ */
+	irq_enable(config->irq.line);
+
+out:
+	irq_unlock(key);
+
+	return r;
+}
+
+static inline void i2c_esp32_ack_intrs(const struct i2c_esp32_config *config)
+{
+	/* Acknowledge all interrupts */
+	*config->regs.intr_clr = ~0;
+}
+
+
+static inline void i2c_esp32_reset_fifo(const struct i2c_esp32_config *config)
+{
+	/* Writing 1 and then 0 to these bits will reset the I2C fifo */
+	*config->regs.fifo_cfg |= (I2C_TX_FIFO_RST | I2C_RX_FIFO_RST);
+	*config->regs.fifo_cfg &= ~(I2C_TX_FIFO_RST | I2C_RX_FIFO_RST);
+}
+
+static inline void i2c_esp32_prepare(const struct i2c_esp32_config *config)
+{
+	i2c_esp32_ack_intrs(config);
+	i2c_esp32_reset_fifo(config);
+}
+
+static int i2c_esp32_spin_yield(int *counter)
+{
+	*counter = *counter + 1;
+
+	if (*counter > I2C_ESP32_TIMEOUT) {
+		return -ETIMEDOUT;
+	}
+
+	if (*counter > I2C_ESP32_SPIN_THRESHOLD) {
+		k_yield();
+	}
+
+	return 0;
+}
+
+static int i2c_esp32_transmit_and_wait(struct device *dev,
+				       volatile struct i2c_esp32_cmd *wait_cmd)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	u32_t status;
+	int counter = 0;
+	int r;
+
+	i2c_esp32_ack_intrs(config);
```

Comment:
This looks like a race.  Are you absolutely, positively sure that before the entry to this function it's not possible even in the abstract for an interrupt to be posted to the device from any source anywhere?  If not, this is going to potentially ACK it before the interrupt is delivered to the CPU.

What's the purpose anyway?  We should be ACKing interrupts inside interrupt handlers, after having handled them.  What does this do?   If you want to be sure the hardware is at a known state, maybe call it "reset" instead?